libdisco/asymmetric.go

```diff
@@ -59,85 +62,173 @@ func dh(keyPair KeyPair, publicKey [32]byte) (shared [32]byte) {
+// Sign a message using a deterministic nonce
+func (kp SigningKeypair) Sign(message []byte) Signature {
+
+	// choose a deterministic nonce
+
+	var kBytes [64]byte
+	// preallocate and avoid append allocations
+	var buf = make([]byte, 0, skSize+len(message))
+	buf = append(buf, kp.SecretKey.Encode(nil)...)
+	buf = append(buf, message...)
+	// buf ownership passed to reusable buffer
+	var reusableBuffer = bytes.NewBuffer(buf)
+	// the output should be 64 byte per ristretto255 rfc
+	// ref : https://tools.ietf.org/html/draft-hdevalence-cfrg-ristretto-00
+	copy(kBytes[:], Hash(reusableBuffer.Bytes(), 64))
+	// cleanup
+	reusableBuffer.Reset()
 	var k ristretto.Scalar
-	k.SetBytes(&kBytes)
-	var r ristretto.Point
-	r.ScalarMultBase(&k)
-
-	// 2. e = H(r || M)
-	e := Hash(append(r.Bytes(), message...), 32)
-	var e32 [32]byte
-	copy(e32[:], e)
-	// 3. s = k - xe
-	var xe ristretto.Scalar
-	var eScal ristretto.Scalar
-	eScal.SetBytes(&e32)
-	xe.Mul(&kp.SecretKey, &eScal)
+	var R ristretto.Element
+
+	k.FromUniformBytes(kBytes[:])
+	R.ScalarBaseMult(&k)
+
+	var e ristretto.Scalar
+	var x ristretto.Scalar
 	var s ristretto.Scalar
-	s.Sub(&k, &xe)
 
-	// 4. return (s, e)
-	return append(s.Bytes(), e...)
+	// e = H(R||message)
+	reusableBuffer.Write(R.Encode(nil))
+	reusableBuffer.Write(message)
+
+	h := Hash(reusableBuffer.Bytes(), 64)
+	e.FromUniformBytes(h)
+
+	reusableBuffer.Reset()
+	// x = sk*e
+	x.Multiply(&kp.SecretKey, &e)
+	// s = k + x
+	s.Add(&k, &x)
+
+	sig := Signature{R, s}
+
+	return sig
+
 }
 
-//
-func (kp SigningKeyPair) Verify(message, signature []byte) error {
-	// 0. (s, e) = sig
-	if len(signature) != 64 {
-		return fmt.Errorf("disco: signature length incorrect")
-	}
-	var s, e [32]byte
-	copy(s[:], signature[:32])
-	copy(e[:], signature[32:])
-
-	// 1. rv = g^s y^e
-	var gs ristretto.Point
-	var sScal ristretto.Scalar
-	sScal.SetBytes(&s)
-	gs.ScalarMultBase(&sScal)
-	var ye ristretto.Point
-	var eScal ristretto.Scalar
-	eScal.SetBytes(&e)
-	ye.PublicScalarMult(&kp.PublicKey, &eScal)
-
-	var rv ristretto.Point
-	rv.Add(&gs, &ye)
-
-	// 2. ev = H(rv || M)
-	ev := Hash(append(rv.Bytes(), message...), 32)
-
-	// 3. ev = e ?
-	if !bytes.Equal(ev, e[:]) {
-		return fmt.Errorf("disco: signature is invalid")
+// Verify a signature
+func (kp SigningKeypair) Verify(message []byte, signature Signature) (bool, error) {
+
+	// Verifying a signature of the form R,s
+	// Decoding the signature
+
+	var R ristretto.Element
+	var s ristretto.Scalar
+
+	R = signature.R
+	s = signature.S
+
+	ev := Hash(append(R.Encode(nil), message...), 64)
+
+	var k ristretto.Scalar
+	k.FromUniformBytes(ev)
+
+	var Rp ristretto.Element
+	Rp.ScalarBaseMult(&s)
+
+	var ky ristretto.Element
+	ky.ScalarMult(&k, &kp.PublicKey)
+
+	Rp.Subtract(&Rp, &ky)
+
+	return Rp.Equal(&R) == 1, nil
+
+}
+
+// newRandomScalar generates a random ristretto scalar using crypto/rand.
+func newRandomScalar() (ristretto.Scalar, error) {
```

Comment:
These two random functions are only used in one place, so it's better to have their logic in the functions themselves.

Also, I would suggest upstreaming these functions to gtank's library, ideally we wouldn't have them in here :)